src/bans.rs

```diff
@@ -0,0 +1,341 @@
+use arc_swap::ArcSwap;
+use log::error;
+use once_cell::sync::Lazy;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::config::get_config;
+use crate::config::Address;
+use tokio::sync::mpsc;
+use tokio::sync::mpsc::error::TrySendError;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::time::Duration;
+use tokio::time::Instant;
+
+pub type GBanList = HashMap<(String, String), HashMap<Address, BanEntry>>;
+
+#[derive(Debug, Clone, Copy)]
+pub enum BanReason {
+    FailedHealthCheck,
+    MessageSendFailed,
+    MessageReceiveFailed,
+    FailedCheckout,
+    StatementTimeout,
+    #[allow(dead_code)]
+    ManualBan,
+}
+#[derive(Debug, Clone)]
+pub struct BanEntry {
+    reason: BanReason,
+    time: Instant,
+    duration: Duration,
+}
+
+#[derive(Debug, Clone)]
+pub enum BanManagerEvent {
+    Ban {
+        address: Address,
+        pool_name: String,
+        username: String,
+        reason: BanReason,
+    },
+    Unban {
+        address: Address,
+        pool_name: String,
+        username: String,
+    },
+    CleanUpBanList,
+}
+
+impl BanEntry {
+    pub fn has_expired(&self) -> bool {
+        return Instant::now().duration_since(self.time) > self.duration;
+    }
+
+    pub fn is_active(&self) -> bool {
+        !self.has_expired()
+    }
+}
+static BANLIST: Lazy<ArcSwap<GBanList>> = Lazy::new(|| ArcSwap::from_pointee(GBanList::default()));
+
+static BAN_REPORTER: Lazy<ArcSwap<BanReporter>> =
+    Lazy::new(|| ArcSwap::from_pointee(BanReporter::default()));
+
+#[derive(Clone, Debug)]
+pub struct BanReporter {
+    channel_to_worker: Sender<BanManagerEvent>,
+}
+
+impl Default for BanReporter {
+    fn default() -> BanReporter {
+        let (channel_to_worker, _rx) = channel(1000);
+        BanReporter { channel_to_worker }
+    }
+}
+
+impl BanReporter {
+    /// Create a new Reporter instance.
+    pub fn new(channel_to_worker: Sender<BanManagerEvent>) -> BanReporter {
+        BanReporter { channel_to_worker }
+    }
+
+    /// Send statistics to the task keeping track of stats.
+    fn send(&self, event: BanManagerEvent) {
+        let result = self.channel_to_worker.try_send(event.clone());
```

Comment:
Let's make this `send` instead of `try_send`. We really don't want to ever lose this message.